upass/__main__.py

```diff
@@ -274,6 +279,70 @@ def copy_selected(self, originator):
         if pathg in self.passwords:
             self.call_pass(originator, (path, True, False))
 
+    def generate_password(self, originator, args=None):
+        if args:
+            path, length, symbols, force = args
+        else:
+            path, length, symbols, force = '', 16, True, False
+        if not path and self.current != '.':
+            path = self.current + '/'
+        self.set_header('GENERATE PASSWORD')
+        self._clear_box()
+        self.generate_password_path = urwid.Edit(("highlight", "{0}{1}".format(GENERATE_PASSWORD_PROMPT, path)))
+        self.generate_password_length = urwid.IntEdit("Length: ", default=length)
+        self.generate_password_symbols = urwid.CheckBox("Symbols? ", state=symbols)
+        self.generate_password_force = urwid.CheckBox("Overwrite existing passwords with that name? ", state=force)
+        self.box.body.append(self.generate_password_path)
+        self.box.body.append(self.generate_password_length)
+        self.box.body.append(self.generate_password_symbols)
+        self.box.body.append(self.generate_password_force)
+        self.box.body.append(ActionButton('GENERATE', self.call_generate))
+        self.box.body.append(BackButton('BACK', self.load_dispatch,
+                                        self.current, self))
+
+    def call_generate(self, originator):
+        gargs = ['pass', 'generate']
+        symbols = self.generate_password_symbols.state
+        force = self.generate_password_force.state
+        if not symbols:
+            gargs.append('-n')
+        if force:
+            gargs.append('-f')
+        path = self.generate_password_path.text[len(GENERATE_PASSWORD_PROMPT):]
+        length = str(self.generate_password_length.value())
+        gargs += [path, length]
+        self._clear_box()
+        self.box.body.append(urwid.AttrMap(
+            urwid.Text('Generating "{0}": (length: {1}, symbols: {2}, overwrite: {3})'.format(path, length,
+                                                                                              symbols, force)),
+            'highlight'))
+        self.loop.draw_screen()
+        p = subprocess.Popen(gargs, stdout=subprocess.PIPE,
+                             stderr=subprocess.PIPE)
+        if path.endswith('/'):
+            errormsg = 'Invalid key path "{0}". Please enter a valid password name.'.format(path)
+        else:
+            try:
+                stdout, errormsg = p.communicate(timeout=1)
```

Comment:
The timeout is too tight and might fail on some systems (it does on mine, for example). Two better solutions:

1. use `p.stdout` / `p.stderr` instead, and check the pass output (by text comparison)
2. use `os.path.exists(path + '.gpg')`